external-crates/move/move-compiler/src/hlir/translate.rs

```diff
@@ -583,782 +597,1088 @@ fn type_(context: &Context, sp!(loc, ty_): N::Type) -> H::Type {
+            if let (Some(lhs), Some(rhs)) = (lhs, rhs) {
+                block.push_back(make_command(eloc, C::Mutate(Box::new(lhs), Box::new(rhs))));
+                false
             } else {
-                MoveOpAnnotation::InferredNoCopy
-            };
-            HE::Move {
-                annotation,
-                var: translate_var(var),
+                true
             }
         }
-        TE::Copy { from_user, var } => HE::Copy {
-            from_user,
-            var: translate_var(var),
-        },
-        TE::BorrowLocal(mut_, v) => HE::BorrowLocal(mut_, translate_var(v)),
 
-        TE::Use(_) => panic!("ICE unexpanded use"),
-        TE::ModuleCall(call) => {
-            let T::ModuleCall {
-                module,
-                name,
-                type_arguments,
-                arguments,
-                parameter_types,
-                acquires,
-            } = *call;
-            let expected_type = H::Type_::from_vec(eloc, single_types(context, parameter_types));
-            let htys = base_types(context, type_arguments);
-            let hargs = exp_list(context, result, Some(&expected_type), *arguments);
-            let call = H::ModuleCall {
-                module,
-                name,
-                type_arguments: htys,
-                arguments: hargs,
-                acquires,
-            };
-            HE::ModuleCall(Box::new(call))
+        // calls might be for effect
+        e_ @ E::ModuleCall(_) | e_ @ E::Builtin(_, _) => {
+            value_statement(context, block, make_exp(e_));
+            false
+        }
+
+        // -----------------------------------------------------------------------------------------
+        // valued expressions -- when these occur in statement position need their children
+        // unravelled to find any embedded, effectful operations. We unravel those and discard the
+        // results. These cases could be synthesized as ignore_and_pop but we avoid them altogether
+        // -----------------------------------------------------------------------------------------
+
+        // FIXME(cgswords): we can't optimize because almost all of these throw. We have to do the
+        // "honest" work here, even though it's thrown away. Consider emitting a warning about
+        // these and/or weaking guarantees in Move 2024.
+        e_ @ E::Vector(_, _, _, _)
+        | e_ @ E::Dereference(_)
+        | e_ @ E::UnaryExp(_, _)
+        | e_ @ E::BinopExp(_, _, _, _)
+        | e_ @ E::Pack(_, _, _, _)
+        | e_ @ E::ExpList(_)
+        | e_ @ E::Borrow(_, _, _)
+        | e_ @ E::TempBorrow(_, _)
+        | e_ @ E::Cast(_, _)
+        | e_ @ E::Annotate(_, _)
+        | e_ @ E::BorrowLocal(_, _)
+        | e_ @ E::Constant(_, _)
+        | e_ @ E::Move { .. }
+        | e_ @ E::Copy { .. }
+        | e_ @ E::Spec(..)
+        | e_ @ E::UnresolvedError => value_statement(context, block, make_exp(e_)),
```

Comment:
Nit (I really think Rust should lint against this IMO)

```suggestion
        e_ @ (E::Vector(_, _, _, _)
        | E::Dereference(_)
        | E::UnaryExp(_, _)
        | E::BinopExp(_, _, _, _)
        | E::Pack(_, _, _, _)
        | E::ExpList(_)
        | E::Borrow(_, _, _)
        | E::TempBorrow(_, _)
        | E::Cast(_, _)
        | E::Annotate(_, _)
        | E::BorrowLocal(_, _)
        | E::Constant(_, _)
        | E::Move { .. }
        | E::Copy { .. }
        | E::Spec(..)
        | E::UnresolvedError) => value_statement(context, block, make_exp(e_)),
```